internal/research/opentelemetry/instrumentation-metrics.md

```diff
@@ -0,0 +1,43 @@
+# OpenTelemetry SDK Usage for Metrics in Golang Applications
+
+OpenTelemetry is a project to handle telemetry data (logs, metrics and traces) in a vendor-neutral way. It provides SDKs to instrument applications, APIs and processing tools (collector).
+
+This document describes our findings using the OpenTelemetry Golang SDK for metrics export in a microservice. To evaluate the functionality of the SDK, a Golang implementation for the reviews service of the [Istio Bookinfo](https://istio.io/latest/docs/examples/bookinfo/) example was extended.
+
+## Metric Types
+
+OpenTelemetry supports metrics of type `counter` (increasing value) , `observer` (gauge value), and `measure`. Measures can be combined using aggregations. A possible aggregation is a histogram. The metric types are represented by the `Int64Counter`, `Int64UpDownCounter`, and `Int64ValueRecorder` types in the Golang SDK.
+
+## Code Instrumentation
+
+To instrument a Golang application, first, a `Meter` object must be created. To use the use the Prometheus protocol, it has to be exposed as an http handler:
+
+```
+exporter, _ := prometheus.InstallNewPipeline(prometheus.Config{})
+http.HandleFunc("/", exporter.ServeHTTP)
+meter := global.Meter("my-service")
+```
+
+The exporter initialization is identical to the [tracing instrumentation](instrumentation-tracing.md) for the OTLP protocol.
+
+Individual code snippets can be instrumented with metrics by first creating a metric:
+
+```
+requestCounter := metric.Must(meter).NewInt64Counter("http.server.requests.total", metric.WithDescription("measures total the number of HTTP requests"))
+```
+
+And modified for example in a HTTP request handler:
+
+```
+h.requestCounter.Add(r.Context(), 1)
+```
```

Comment:
Imho, would be also nice to document:
* Asynchronous instrumentation (since it's comparable to the asynchronous API of the Prometheus client library): https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-instrument-details
* Synchronous and asynchronous batching: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#recordbatch-calling-convention and https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#batch-observer

Reply:
Done